mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java

```diff
@@ -281,16 +279,63 @@ public MessageMetaData add(Mailbox mailbox, MailboxMessage message) throws Mailb
+        if (!finalResult.getFailed().isEmpty()) {
+            LOGGER.error("Can not update following UIDs {} for mailbox {}", finalResult.getFailed(), mailboxId.asUuid());
+        }
+        return finalResult.getSucceeded().iterator();
+    }
+
+    private FlagsUpdateStageResult handleUpdatesStagedRetry(CassandraId mailboxId, FlagsUpdateCalculator flagUpdateCalculator, FlagsUpdateStageResult firstResult) {
+        FlagsUpdateStageResult globalResult = firstResult;
+        int retryCount = 0;
+        while (retryCount < maxRetries && !globalResult.getFailed().isEmpty()) {
+            retryCount++;
+            globalResult = globalResult.keepSucceded()
+                .merge(retryUpdatesStage(mailboxId, flagUpdateCalculator, globalResult.getFailed()));
+        }
+        return globalResult;
+    }
+
+    private FlagsUpdateStageResult retryUpdatesStage(CassandraId mailboxId, FlagsUpdateCalculator flagsUpdateCalculator, List<MessageUid> failed) {
+        Stream<ComposedMessageIdWithMetaData> idsFailed = FluentFutureStream.of(
+            failed.stream().map(uid -> messageIdDAO.retrieve(mailboxId, uid)))
+            .flatMap(OptionalConverter::toStream)
+            .join();
+
+        return runUpdateStage(mailboxId, idsFailed, flagsUpdateCalculator);
+    }
+
+    private FlagsUpdateStageResult runUpdateStage(CassandraId mailboxId, Stream<ComposedMessageIdWithMetaData> toBeUpdated, FlagsUpdateCalculator flagsUpdateCalculator) {
+        Long newModSeq = modSeqProvider.nextModSeq(mailboxId).join().orElseThrow(() -> new RuntimeException("ModSeq generation failed for mailbox " + mailboxId.asUuid()));
+
+        return toBeUpdated.collect(JamesCollectors.chunker(UPDATE_FLAGS_BATCH_SIZE))
+            .map(uidChunk -> performUpdatesForChunk(mailboxId, flagsUpdateCalculator, newModSeq, uidChunk))
+            .map(CompletableFuture::join)
+            .reduce(FlagsUpdateStageResult.none(), FlagsUpdateStageResult::merge);
+    }
+
+    private CompletableFuture<FlagsUpdateStageResult> performUpdatesForChunk(CassandraId mailboxId, FlagsUpdateCalculator flagsUpdateCalculator, Long newModSeq, Collection<ComposedMessageIdWithMetaData> uidChunk) {
+        Stream<CompletableFuture<FlagsUpdateStageResult>> updateMetaDataFuture =
+            uidChunk.stream().map(oldMetadata -> tryFlagsUpdate(flagsUpdateCalculator, newModSeq, oldMetadata));
+
+        return FluentFutureStream.of(updateMetaDataFuture)
+            .reduce(FlagsUpdateStageResult.none(), FlagsUpdateStageResult::merge)
+            .thenCompose(result -> updateIndexesForUpdatesResult(mailboxId, result));
+    }
+
+    private CompletableFuture<FlagsUpdateStageResult> updateIndexesForUpdatesResult(CassandraId mailboxId, FlagsUpdateStageResult result) {
+        return FluentFutureStream.of(
+            result.getSucceeded().stream()
+                .map((UpdatedFlags updatedFlags) -> indexTableHandler.updateIndexOnFlagsUpdate(mailboxId, updatedFlags)))
```

Comment:
1. They are not expected to fail because we don't use lightweight conditions here

2. The entire update fails as a Runtime exception is thrown. This might lead to index inconsistencies, that we don't repair yet.

But I don't see how it is different from old code behaviour. 

I think table healing should be a topic on its own. I don't think we can tacle the issue here.

About general failure: Would you like me to LOG and Swallow index update fail ? to not abort the entire result?